common/setups/hybrid/util.py

```diff
@@ -376,34 +407,81 @@ def __init__(
             'alignment': "train_vtln",  # if using run function not needed
         }
         ##################################################
-        :param vtln_sat_recognition_args:
+        :param recognition_args:
         ##################################################
         """
-        self.vtln_sat_training_args = vtln_sat_training_args
-        self.vtln_sat_recognition_args = vtln_sat_recognition_args
-        self.sdm_vtln_sat_args = sdm_vtln_sat_args
+        self.training_args = training_args
+        self.recognition_args = recognition_args
+        self.sdm_args = sdm_args
+
+
+class ForcedAlignmentArgs:
+    def __init__(self, name, target_corpus, flow, feature_scorer):
+        self.name = name
+        self.target_corpus = target_corpus
+        self.flow = flow
+        self.feature_scorer = feature_scorer
+
+
+class ReturnnRasrDataInput(RasrDataInput):
+    def __init__(
+        self,
+        segment_path: Optional[Union[tk.Path, str]],
+        cart_tree: Optional[Union[tk.Path, str]],
+        alignments: Optional[
+            Union[tk.Path, str, MultiPath, rasr.FlagDependentFlowAttribute]
+        ],
+        features: Optional[
+            Union[tk.Path, str, MultiPath, rasr.FlagDependentFlowAttribute]
+        ],
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.segment_path = segment_path
+        self.cart_tree = cart_tree
+        self.alignments = alignments
+        self.features = features
 
 
 class NnArgs:
     def __init__(
         self,
-        nn_training_args: Optional[dict] = None,
-        nn_count_recognition_args: Optional[dict] = None,
-        nn_nn_recognition_args: Optional[dict] = None,
-        nn_rescoring_args: Optional[dict] = None,
+        returnn_configs: Dict[str, returnn.ReturnnConfig],
+        training_args: Optional[dict] = None,
+        count_recognition_args: Optional[dict] = None,
+        lm_recognition_args: Optional[dict] = None,
```

Comment:
> both are lm_recognition_args, I suggest to rename lm_recognition_args to neural_recognition_args.

yes

> But then why make this distinction? we can have a list/dict/tuple of recognition_args and then some of them might have a neural lm configured and some have a count lm configured.

My reasoning was that normally we want to do count and neural recognition for a given model. That is why I wanted to have both. But not sure if this applies to all users..

Reply:
Some users would only do count based LM recognitions, other users might want to do [count_2gram, count_4gram, ffnn_4gram, lstm_small, lstm_big, trafo] recognitions. 

As far as I can see, there would be no special different handling of count vs. neural recognition args except maybe for a different alias.

I would implement it e.g. as a dict of dicts with `lm_name` -> `recognition_args_dict` so that we can have an arbitrary number of recognition settings.

Reply:
OK. Agreed.